src/vm/methodtable.inl

```diff
@@ -175,6 +175,13 @@ inline PTR_Module MethodTable::GetLoaderModule()
     return ReadPointer(this, &MethodTable::m_pLoaderModule);
 }
 
+//==========================================================================================
+inline PTR_Module MethodTable::GetLoaderModuleIfPossible()
+{
+    LIMITED_METHOD_DAC_CONTRACT;
+    return m_pLoaderModule.IsNull() ? NULL : ReadPointer(this, &MethodTable::m_pLoaderModule);
```

Comment:
LoaderModule is never null in correctly constructed MethodTable. Could you please figure out and fix the root cause why you are getting invalid MethodTable with null LoaderModule instead of adding the workaround?

Reply:
I did do some investigation to see who is creating these objects, but it happens pretty early in the startup and so `ObjectAllocated` isn't called for these objects. All these objects are either minimum object size (24 bytes) or 32-bytes. I haven't seen any that aren't those.

This behavior also occurs in full framework.

Reply:
It looks like you are hitting fake padding objects with `g_pFreeObjectMethodTable`. These objects are meant to be internal to the GC. They are not meant to escape to the rest of the system. It is why you are do not see `ObjectAllocated` to be called for them.

What is the API that is giving you these objects back? Would it be better to filter them out there?

Reply:
I investigated this and these are objects created by the GC and have always been there. What's really surprising is that no one ever hit this, but now that we've made GC profiling usable and people will want to use it we should fix this.

Reply:
> It looks like you are hitting fake padding objects with `g_pFreeObjectMethodTable`. These objects are meant to be internal to the GC. They are not meant to escape to the rest of the system. It is why you are do not see `ObjectAllocated` to be called for them.
> 
> What is the API that is giving you these objects back? Would it be better to filter them out there?

Oops, GitHub didn't show me this comment.

[ICorProfilerInfo2::GetGenerationBounds](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilerinfo2-getgenerationbounds-method)

I haven't looked but it would seem somewhat expensive to filter this, or create subranges in the segments returned by that API.

The other option could be to create a valid MT if this change is not acceptable?

Reply:
This API returns memory ranges. How do you go from memory range to the object pointer?

Reply:
Yes it returns ranges, but each range starts from an object pointer.

Reply:
Hopefully the change now is acceptable.

Reply:
> it returns ranges, but each range starts from an object pointer.

Is that pointer useful when it is just the first object pointer? How can you get to the next object?

It looks more like a bug in the API prototype. The range start should have been `void*`.

Reply:
The API prototype is quite alright actually.

It returns an `ObjectID` which is used throughout the Profiler API to mean an object pointer. You then call [ICorProfilerInfo4::GetObjectSize2](https://docs.microsoft.com/en-us/dotnet/framework/unmanaged-api/profiling/icorprofilerinfo4-getobjectsize2-method) to skip the object. Then you align it to `SIZE_T`. One exception to that is on x86 you align to 8-bytes when iterating a LOH segment.

Reply:
@Maoni0 Do we assume this heap walking algorithm to be a public contact?

Reply:
https://github.com/dotnet/coreclr/blob/master/Documentation/Profiling/davbr-blog-archive/GC%20Heap%20and%20Alignment%20Padding.md

Reply:
It is also on MSDN https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/ms231885(v=vs.100)

> The size returned by the GetObjectSize method does not include any alignment padding that may appear after the object is on the garbage collection heap. If you use the GetObjectSize method to advance from object to object on the garbage collection heap, add alignment padding manually, as necessary.

> On 32-bit Windows, COR_PRF_GC_GEN_0, COR_PRF_GC_GEN_1, and COR_PRF_GC_GEN_2 use 4-byte alignment, and COR_PRF_GC_LARGE_OBJECT_HEAP uses 8-byte alignment.

> On 64-bit Windows, the alignment is always 8 bytes.

Reply:
Ok. I would expect these details to be hidden behind API, but since we have documented via a blog post, it is too late to really fix it.